lib/jnpr/junos/utils/start_shell.py

```diff
@@ -112,10 +124,11 @@ def run(self, command, this=_SHELL_PROMPT):
         # run the command and capture the output
         self.send(command)
         got = ''.join(self.wait_for(this))
-
-        # use $? to get the exit code of the command
-        self.send('echo $?')
-        rc = ''.join(self.wait_for(this))
+        rc = ''
+        if re.search(r'(%|>|#)\s?$', got) is not None:
+            # use $? to get the exit code of the command
+            self.send('echo $?')
+            rc = ''.join(self.wait_for(this))
         self.last_ok = True if rc.find('0') > 0 else False
```

Comment:
What should the value of `self.last_ok` be when `this != _SHELL_PROMPT` (looking for a non-default prompt)? With the current code, it is set to `False`.

When `this != _SHELL_PROMPT` I believe we should instead set `self.last_ok` to `True` if `this` was found and `False` if `this` was not found.

See the comment above on line 127.

If you agree with my approach, then this line should be part of the conditional block above it.